Automatically create release tags via GitHub actions

tautschnig opened the pull request:
~The action is currently marked as DRY_RUN: true to perform testing, and
will be enabled in a subsequent PR.~ Edit: I have performed testing in my fork (see https://github.com/tautschnig/cbmc/runs/1533820084), and am thus inclined to immediately enable this over here.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Automatically create release tags via GitHub actions

Release tags are now created automatically when including an indicator
(see updated documentation) in the commit message.

hannes-steffenhagen-diffblue commented on .github/workflows/create-release-tag.yaml:
```diff
@@ -0,0 +1,19 @@
+    - name: Bump version and push tag
+      uses: anothrNick/github-tag-action@1.33.0
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Have you checked if the default `GITHUB_TOKEN` has the rights to create a tag? It not being allowed to do some things is something that has bitten me in the past, that's why I’m asking.

tautschnig replied:
Good call out. I don't know whether this works, but also I'm not sure how to figure it out other than trying it out? (It seems safe to do so: in the worst case, it does not create a tag, and we'd have to go back to the previous, manual approach.)

piotr-grabalski commented on .github/workflows/create-release-tag.yaml:
```diff
@@ -0,0 +1,19 @@
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        DEFAULT_BUMP: none
+        RELEASE_BRANCHES: develop
+        VERBOSE: false
```
Does this read the latest version from the existing tags?
If not we may need INITIAL_VERSION too:
> INITIAL_VERSION (optional) - Set initial version before bump. Default 0.0.0.

Probably something like: `INITIAL_VERSION: 5.20.1`?

tautschnig replied:
Supposedly the bash script behind this GitHub action does check for existing tags. But it may be worth flipping `DRY_RUN` back on to make sure this does actually work as expected.

tautschnig replied:
I have now added a commit to keep in in `DRY_RUN` mode.

tautschnig commented:
A further comment: we should likely wait for https://github.com/anothrNick/github-tag-action/pull/95 to be merged to maintain the `cbmc-` prefix.

tautschnig committed:
Keep automated release tagging in dry-run mode

We should first make sure this actually works on this repository before
enabling it in production.

NlightNFotis commented:
Hi Michael, I'd love to get this in and experiment with this a bit - it would save us quite a bit of time!

Are you happy to take this forward (rebase on `develop`) or if you don't have the bandwidth for this, are you happy with me taking this over?*

*: (I'll close this PR and make a new one, but `cherry-pick` your commits on top of the new branch, and any changes by me will go in another commit, so your authorship is going to be preserved)

tautschnig commented:
Closing: automatic tagging won't save much time as the tag can be produced at the same time as publishing the changelog.
